cpp/ql/src/Security/CWE/CWE-120/OverrunWrite.ql

```diff
@@ -21,14 +21,14 @@ import semmle.code.cpp.commons.Alloc
  * See CWE-120/UnboundedWrite.ql for a summary of CWE-120 alert cases.
  */
 
-from BufferWrite bw, Expr dest, int destSize
+from BufferWrite bw, Expr dest, int destSize, BufferWriteEstimationReason reason
 where
   not bw.hasExplicitLimit() and // has no explicit size limit
   dest = bw.getDest() and
   destSize = getBufferSize(dest, _) and
   // we can deduce that too much data may be copied (even without
   // long '%f' conversions)
-  bw.getMaxDataLimited() > destSize
+  bw.getMaxDataLimited(reason) > destSize
 select bw,
   "This '" + bw.getBWDesc() + "' operation requires " + bw.getMaxData() +
-    " bytes but the destination is only " + destSize + " bytes."
+    " bytes but the destination is only " + destSize + " bytes (" + reason.toString() + ")."
```

Comment:
Do we really want to write out this `reason` in the description? I'm guessing users don't really care if we obtained a bound using type-based bounds or range analysis-bounds?

Reply:
well, the range analysis case will be moved to a separate query, but the plan is to have the float case here as well, then it will probably be good to point that out. But we could be less wordy when we get to that

Reply:
Why would it be important for the float case?

Reply:
I think for the same reason that we created the special float case query in the first place. Many might not know that a `%f` might have such extreme conversion lengths, so when they see that a format string containing "%f" is reported to potentially be hundreds of bytes long they might get surprised. I know I did not know that before (lucky me, I actually never needed to program with floating point numbers, ever :sweat_smile:).

Having something in the query result explicitly pointing to that fact might clear it up instantly.

Reply:
I might be misunderstanding, but isn't the text in parentheses either:
- "based on type bounds", or
- "based on flow analysis of value bounds"
So is the idea that when programmers see the string "(based on type bounds)", they will know that we deduced a potential overflow due to floating-point specifiers?

Reply:
sorry, I think I confused you :smile: Currently no, floating point estimates are not included in this query. I was speaking about plans for the near future, when the `based on flow analysis` reason will not be visible in this query as it will be moved to another high precision query, but we will then move here the float case. We can maybe then tweak the output then, when we cross that bridge.

So, I can remove this altogether now. It was quite useful to check the tests now, to look whether the reason was the correct one without changing the query results just yet. And also, it helped out spot one case in an earlier stage of development that was giving multiple reasons (without explicitly using the reason in the `select` part this was hidden in the output), but it's not really needed before the next steps.

Reply:
(which by the way will make this PR independent of the internal one)

Reply:
I see. Thanks for clearing up my confusion. Yeah, I think it's best to avoid this line currently. We might manage to do a dist upgrade on LGTM before the high-precision query work is done, and in that case, I don't think the message will be useful to programmers.